File: problems/easy/589.n_ary_tree_preorder_traversal.cpp
#include <vector>
#include <stack>
using namespace std;
/*
    [Problem]
        - Given the root of an n-ary tree, return the preorder traversal of its nodes' values.
          Preorder traversal means : ROOT -> LEFT CHILD -> RIGHT CHILD .
          ( Inorder : LEFT -> ROOT -> RIGHT, 
            Postorder : LEFT -> RIGHT -> ROOT )

*   [Idea]
        - Use Stack (implement by Stack pop(), push() in C++) to store nodes of the given n-ary tree.
        - Use stack.front() to get the top node and pop() the node from stack (Last In First Out manner),
          push children nodes of the current node into the stack in reversed order. (Since pre-order traversal)
        
    [Complexity]
        - Time : O(N), where N is the number of nodes in the given n-ary tree.
        - Space : O(N), use stack (implemented by Stack in C++) for extra space.
*/

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};


class Solution {
public:
    vector<int> preorder(Node* root) {
        stack<Node*> stack_;
        vector<int> result;
        if(!root) return result;
        stack_.push(root);

        while(!stack_.empty()){
            Node* curr = stack_.top();
            stack_.pop();
            result.push_back(curr->val);

            int childrenNum = curr->children.size();
            while(childrenNum -- ){
                stack_.push(curr->children[childrenNum]);
            }
        }
        return result;
    }
};

File: problems/easy/589.n_ary_tree_preorder_traversal.py
"""  [Problem]
        - Given the root of an n-ary tree, return the preorder traversal of its nodes' values.
          Preorder traversal means : ROOT -> LEFT CHILD -> RIGHT CHILD .
          ( Inorder : LEFT -> ROOT -> RIGHT, 
            Postorder : LEFT -> RIGHT -> ROOT )

*   [Idea]
        - Use DFS and Stack (implement by List pop(), append() in Python) to store nodes of the given n-ary tree.
        - Pop the node from stack (Last In First Out manner) and extend reversed order of its children into stack,
          so it could pop() the last item from stack in the next iteration.
        
    [Complexity]
        - Time : O(N), where N is the number of nodes in the given n-ary tree.
        - Space : O(N), use stack (implemented by List in Python) for extra space.
"""

"""
# Definition for a Node.
"""
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """

        if root is None : return []
        # use List in Python to implement stack (LIFO), if implementing queue, use deque instead(append() and popleft())
        stack = [root] 
        output = [] 
        
        while stack : 
            curr = stack.pop()
            output.append(curr.val)
            stack.extend(curr.children[:: -1 ])
        
        return output